frontend/src/components/widgets/Radio/Radio.tsx

```diff
@@ -14,128 +14,197 @@
-import { FormClearHelper } from "src/components/widgets/Form"
-import { WidgetStateManager, Source } from "src/lib/WidgetStateManager"
+import { WidgetStateManager } from "src/lib/WidgetStateManager"
 import { labelVisibilityProtoValueToEnum } from "src/lib/utils"
 
 export interface Props {
   disabled: boolean
+  theme: Theme
   element: RadioProto
   widgetMgr: WidgetStateManager
   width: number
 }
 
-interface State {
-  /**
-   * The value specified by the user via the UI. If the user didn't touch this
-   * widget's UI, the default value is used.
-   */
-  value: number
-}
-
-class Radio extends React.PureComponent<Props, State> {
-  private readonly formClearHelper = new FormClearHelper()
+function Radio(props: Props): ReactElement {
+  const { theme, element, width, widgetMgr } = props
+  let { disabled } = props
+  const { colors, radii } = theme
+  const style = { width }
+  const { horizontal, options, label, labelVisibility, help } = element
 
-  public state: State = {
-    value: this.initialValue,
-  }
-
-  get initialValue(): number {
+  const initialValue = (): number => {
     // If WidgetStateManager knew a value for this widget, initialize to that.
     // Otherwise, use the default value from the widget protobuf.
-    const storedValue = this.props.widgetMgr.getIntValue(this.props.element)
-    return storedValue !== undefined ? storedValue : this.props.element.default
-  }
-
-  public componentDidMount(): void {
-    if (this.props.element.setValue) {
-      this.updateFromProtobuf()
-    } else {
-      this.commitWidgetValue({ fromUi: false })
-    }
+    const storedValue = widgetMgr.getIntValue(element)
+    return storedValue !== undefined ? storedValue : element.default
   }
 
-  public componentDidUpdate(): void {
-    this.maybeUpdateFromProtobuf()
-  }
+  // Tracks value specified by user via UI
+  const [value, setStateValue] = useState(initialValue())
+  // Tracks whether the value was changed by the user
+  const [source, setStateSource] = useState({ fromUi: false })
 
-  public componentWillUnmount(): void {
-    this.formClearHelper.disconnect()
+  const updateFromProtobuf = (): void => {
+    element.setValue = false
+    setStateValue(element.value)
+    setStateSource({ fromUi: false })
   }
 
-  private maybeUpdateFromProtobuf(): void {
-    const { setValue } = this.props.element
+  const maybeUpdateFromProtobuf = (): void => {
+    const { setValue } = element
     if (setValue) {
-      this.updateFromProtobuf()
+      updateFromProtobuf()
     }
   }
 
-  private updateFromProtobuf(): void {
-    const { value } = this.props.element
-    this.props.element.setValue = false
-    this.setState({ value }, () => {
-      this.commitWidgetValue({ fromUi: false })
-    })
+  const checkNoOptions = (): void => {
+    if (options.length === 0) {
+      options.push("No options to select.")
+    }
+    if (options[0] === "No options to select.") {
+      disabled = true
+    }
   }
```

Comment:
A couple things here:

* This mutates the input argument, which is mildly illegal (the changes are visible to any code outside of this component that may be able to read `element.options`). The old implementation got around this by cloning the options array, but I think we can do better here (more on that below)
* I think that having `disabled` be mutable and potentially changed from this function isn't super great even if it doesn't mutate the input (since `disabled` is a primitive instead of a reference) just from a style point of view. It's not as bad in class-based code, but `let` should ideally be avoided when possible in more functional code (just for sake of convention)
* We may want to just inline this since I don't think having this live in a helper function is much more readable as it's only called in one place

As for what to do about `options`/`disabled`. It may make sense to extract these fields from `element`/`props` as `rawOptions`/`externallyDisabled` (I didn't really put much thought into these names, so feel free to choose better ones), then write something like

```typescript
const disabled = externallyDisabled || rawOptions.length === 0
const options = rawOptions.length === 0 ? ["No options to select."] : rawOptions
```

which should be equivalent but avoids mutating the input or even mutating a local variable.